adds a `run_examples_on_click` parameter to `gr.ChatInterface` mirroring the the `run_on_click` parameter in `gr.Examples`

abidlabs opened the pull request:
Small PR that adds a `run_examples_on_click` parameter on to `gr.ChatInterface`, closes: https://github.com/gradio-app/gradio/issues/10103

Two examples to test with:

```py
import gradio as gr

def echo(message, history):
    return message["text"]

demo = gr.ChatInterface(
    fn=echo,
    type="messages",
    examples=[{"text": "abc"}, {"text": "hello", "files": ["files/avatar.png"]}, {"text": "hola", "files": ["files/cantina.wav"]}],
    title="Echo Bot",
    multimodal=True,
    run_examples_on_click=False,
)
demo.launch()
```

```py
import random
import gradio as gr

def random_response(message, history):
    return random.choice(["Yes", "No"])

demo = gr.ChatInterface(random_response, type="messages", autofocus=False, run_examples_on_click=False, examples=["Hello", "Hi", "How are you?", "I am fine", "Thank you"])

if __name__ == "__main__":
    demo.launch()
```

Also fixes: https://github.com/gradio-app/gradio/issues/10105

freddyaboulton commented:
If `run_on_click=True`, then images in the multimodal textbox are not displayed as images in the chatbot:

![run_on_click_true_multimodal](https://github.com/user-attachments/assets/dce07c1d-8d04-4bcc-a67b-d0ef692373e3)

This is not the case when `run_on_click=False`
![run_on_click_true_multimodal](https://github.com/user-attachments/assets/51728c1e-18e1-47d9-8ad6-c04ad4fbf3cf)'

other than that LGTM!

abidlabs commented:
Thanks yet again @freddyaboulton! Will take a look at that issue

abidlabs commented:
Oh so the issue is actually with the default behavior hmm

abidlabs commented:
Noticed another issue, which is that when `cache_examples=True`, any cached files appear on the wrong side of the chatbot: 

<img width="1264" alt="image" src="https://github.com/user-attachments/assets/bc6a175f-0424-4f00-a63d-86eb4f935df4">

```py
import gradio as gr

def echo(message, history):
    return message["text"]

demo = gr.ChatInterface(
    fn=echo,
    type="messages",
    examples=[{"text": "hello", "files":["files/avatar.png"]}, {"text": "hola"}, {"text": "merhaba"}],
    cache_examples=True,
    title="Echo Bot",
    multimodal=True,
)
demo.launch()
```

abidlabs commented on gradio/chat_interface.py:
```diff
@@ -606,11 +611,8 @@ def _append_multimodal_history(
-                    )
-                else:
-                    history.append({"role": "user", "content": (x,)})  # type: ignore
+                    x = x.get("path")
```
Files should always be stored as tuples in the history, for consistency and as per the documentation.
<img width="867" alt="image" src="https://github.com/user-attachments/assets/e33a61ed-e8fb-465b-8c52-9c9aa8e56f62">

abidlabs commented on gradio/chat_interface.py:
```diff
@@ -788,16 +800,18 @@ def _process_example(
-                    result.append([message["text"], None])
                 for file in message.get("files", []):
                     result.append([file, None])
+                if "text" in message:
```
For consistency with the non-caching examples case, we should add files to the history first and then the files.

abidlabs commented on gradio/components/chatbot.py:
```diff
@@ -507,6 +508,7 @@ def _postprocess_messages_tuples(self, value: TupleFormat) -> ChatbotDataTuples:
     def _postprocess_message_messages(
         self, message: MessageDict | ChatMessage
     ) -> Message:
+        message = copy.deepcopy(message)
```
This is needed because if you yield a sequence of messages into Chatbot and they include files, they will be mutated in-place. This was happening in `gr.ChatInterface` here:

```py
    def example_clicked(
        self, example: SelectData
    ) -> Generator[
        tuple[TupleFormat | list[MessageDict], str | MultimodalPostprocess], None, None
    ]:
        """
        When an example is clicked, the chat history (and saved input) is initially set only
        to the example message. Then, if example caching is enabled, the cached response is loaded
        and added to the chat history as well.
        """
        history = []
        self._append_multimodal_history(example.value, None, history)
        example = self._flatten_example_files(example)
        message = example.value if self.multimodal else example.value["text"]
        yield history, message
        if self.cache_examples:
            history = self.examples_handler.load_from_cache(example.index)[0].root
            yield history, message

```

abidlabs commented:
I think I got everything, if you could give it another look @freddyaboulton @hysts

hysts commented:
Thanks! Basically LGTM, but I noticed that the undo button is not working correctly for multimodal examples `run_examples_on_click=True`. Looks like it was broken from the beginning.

https://github.com/user-attachments/assets/b905722d-2c75-4208-8c86-ce3b6d5f7200

abidlabs commented:
Dang okay thanks for flagging!

hysts reviewed (approved):
As the undo issue is not introduced in this PR, I guess it can be fixed in another PR. So I'll approve this PR for now.

abidlabs commented:
Let me see if I can fix it real quick

abidlabs commented:
Fixed the undo bug and added two more end-to-end tests to test. We've added quite a few e2e tests for chatinterface, but I think we need them all since there are so many different message types and configuration paths. When we have bandwidth, we can figure out ways to run them faster.

abidlabs commented:
Thanks @freddyaboulton and @hysts for the thorough reviews, I'll merge this in when CI is green.

abidlabs commented:
😅 taking a bit longer than I thought but doing some secondary cleanup as well

abidlabs commented:
Whew finally ready. Will merge this in tomorrow if there are no objections!
